tests/test_compiler.py

```diff
@@ -206,3 +207,206 @@ def circuit(x: float):
+                qml.Hadamard(wires=i)
+
+            # QFT
+            @qml.for_loop(0, n, 1)
+            def qft(i):
+                qml.Hadamard(wires=i)
+
+                @qml.for_loop(i + 1, n, 1)
+                def inner(j):
+                    qml.ControlledPhaseShift(np.pi / 2 ** (n - j + 1), [i, j])
+
+                inner()
+
+            init()
+            qft()
+
+            # Expected output: |100...>
+            return qml.state()
+
+        assert jnp.allclose(circuit(4), jnp.eye(2**4)[0])
+
+    def test_cond(self):
+        """Test condition with simple true_fn"""
+        dev = qml.device("lightning.qubit", wires=1)
+
+        @qml.qjit
+        @qml.qnode(dev)
+        def circuit(x: float):
+            def ansatz_true():
+                qml.RX(x, wires=0)
+                qml.Hadamard(wires=0)
+
+            qml.cond(x > 1.4, ansatz_true)
+
+            return qml.expval(qml.PauliZ(0))
+
+        assert jnp.allclose(circuit(1.4), 1.0)
+        assert jnp.allclose(circuit(1.6), 0.0)
+
+    def test_cond_with_else(self):
+        """Test condition with simple true_fn and false_fn"""
+        dev = qml.device("lightning.qubit", wires=1)
+
+        @qml.qjit
+        @qml.qnode(dev)
+        def circuit(x: float):
+            def ansatz_true():
+                qml.RX(x, wires=0)
+                qml.Hadamard(wires=0)
+
+            def ansatz_false():
+                qml.RY(x, wires=0)
+
+            qml.cond(x > 1.4, ansatz_true, ansatz_false)
+
+            return qml.expval(qml.PauliZ(0))
+
+        assert jnp.allclose(circuit(1.4), 0.16996714)
+        assert jnp.allclose(circuit(1.6), 0.0)
+
+    def test_cond_with_elif(self):
+        """Test condition with a simple elif branch"""
+        dev = qml.device("lightning.qubit", wires=1)
+
+        @qml.qjit
+        @qml.qnode(dev)
+        def circuit(x):
+            def true_fn():
+                qml.RX(x, wires=0)
+
+            def elif_fn():
+                qml.RY(x, wires=0)
+
+            def false_fn():
+                qml.RX(x**2, wires=0)
+
+            qml.cond(x > 2.7, true_fn, false_fn, ((x > 1.4, elif_fn),))
+
+            return qml.expval(qml.PauliZ(0))
+
+        assert jnp.allclose(circuit(1.2), 0.13042371)
+        assert jnp.allclose(circuit(jnp.pi), -1.0)
+
+    def test_cond_with_elifs(self):
+        """Test condition with multiple elif branches"""
+        dev = qml.device("lightning.qubit", wires=1)
+
+        @qml.qjit
+        @qml.qnode(dev)
+        def circuit(x):
+            def true_fn():
+                qml.RX(x, wires=0)
+
+            def elif1_fn():
+                qml.RY(x, wires=0)
+
+            def elif2_fn():
+                qml.RZ(x, wires=0)
+
+            def false_fn():
+                qml.RX(x**2, wires=0)
+
+            qml.cond(x > 2.7, true_fn, false_fn, ((x > 2.4, elif1_fn), (x > 1.4, elif2_fn)))
+
+            return qml.expval(qml.PauliZ(0))
+
+        assert jnp.allclose(circuit(1.5), 1.0)
```

Comment:
I didn't expect this. I thought it would do the following:
1. x is not greater than 2.7, so do `false_fn` (`qml.RX(2.25, 0)`)
2. x is not greater than 2.4, skip
3. x is greater than 1.4, do `elif2_fn` (`qml.RZ(1.5, 0)`, no effect on result)
4. Get the result (array(-0.62817362))

I'm now seeing that the `elifs` occur before the `false_fn` which surprised me, but even considering that I didn't expect this result

As well, along with my previous comment on `elifs` all being applied, can you add a test with something like `x=2.5` (between 2.4 and 2.7 so it fails the condition, but passes all elif conditions)

Reply:
looking back, I'm guessing it does what one might naturally assume it does:
1. x > 2.7? no
2. x > 2.4? no
3. x > 1.4? yes - `qml.RZ(1.5, 0)` (which has no effect, result is 1.0)
4. don't do `false_fn` because you did `elif2_fn` already

Reply:
The semantics of `catalyst.cond` are meant to emulate that of Python's (and other languages') conditionals. So just like
```py
if x > 2.7:
  qml.RX(x, wires=0)
elif x > 2.4:
  qml.RY(x, wires=0)
elif x > 1.4:
  qml.RZ(x, wires=0)
else:
  qml.RX(x**2, wires=0)
```
will execute the first branch that is true, and none of the others, so should `catalyst.cond`.

Admittedly, the `qml.cond` syntax makes this harder to see, since it doesn't have any of the structure the Python version does, but the meaning should be the same.

Do you think there is a syntax for `qml.cond` that would be clearer, while maintaining compatibility with the existing version?

Reply:
I'm not sure myself, but I definitely felt that the `elifs` were rather disjointed from the `true_fn` and `false_fn`. However, that might just be the bias of me being familiar with `qml.cond` as it was before these proposed changes. It's probably fine, but if my slowness at grasping what was happening worried you, maybe a lil compare-to-interpreted-python example would be nice? eg. you could re-write the example in the cond docs saying something like "this is equivalent to this interpreted python code:"
```python
@qml.qnode(dev)
def circuit(x):

    if x > 2.7:
        qml.RX(x, wires=0)

    elif x > 1.4:
        qml.RY(x, wires=0)

    else:
        qml.RX(x ** 2, wires=0)

    return qml.expval(qml.PauliZ(0))
```